### ISSUE TITLE
Bugs, Features, Usability problems

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#2236]
I'd like to be sure we're on the same page for tracking and why.

First, context and definitions, I've been thinking of the three terms as:
- bug: something is not working correctly
- usability problem: it works, but the feature is arguably broken, it's hard to use. If we were to test the feature today we would not accept the functionality, it is incomplete or confusing. 
- feature request: this is 'wish list' thing, something brand new, not necessarily any improvement but something new to add to the game. 

We decided to move feature requests to forum for a couple of reasons:
1. community engagement on the topic
2. our bug queue is a task queue, we do not want to track things in that queue that are going to sit around and linger. Wish list items are valuable to track, but we do not want those items clouding up the things that are broken or not working well. It's also fundamentally slightly different tasks for a developer to do a bug fix or a UX improvement, compared to building something new.

Bugs and usability problems we could track in forums, I raised the topic and it seemed not have been a popular proposal. Forums I do believe are weak when dealing with a high volume of issues, they are good for long conversational threads, not detailed task tracking (at least I have not seen how that would work well).

The above definitions do seem clear to me. If a usability problem is really hard to fix and is very low value, and is just sitting around for a long time, and likely to be casually fixed when we do something else, then it's probably a good thing to "ice box". I'm not sure if we gain anything at all by moving something so it can then just sit in the forums. In general I think we should stop the practice of asking for something to be moved and just ice-box stuff here in-place. If bugs are opened in forums, then let's just tag it and ask that bugs be created next time in github (or maybe not even bother since long term we clearly want bug reporting to be more built in).

--------------------------------

So mainly I want to clear up the definitions and expectations. We can certainly choose to change how we label things and how we track, it is open for debate! Though, we should balance our efficiency and what we need vs the pain of reporting problems. My philosophy is basically things that "suck" about TripleA, whether broken or just seemingly broken, should be tracked and are important to fix because we do not want TripleA to suck. On the other hand, idea and wish list are very useful things, for a number of reasons, though as established we do not want to track those items with everything else. What we track in github is ideally actionable, a developer can pick up an issue, read it and after 5 minutes know what to do and can start work. Usability problems are tricky, likely a proposed fix would have to be opened in forums or github issues first, or maybe just floated in a PR with risk that it may be rejected.

**Comment by ron-murhammer** [comment]
Agree with trying to get more clarification. Given your 3 categories, I tend to group 'feature request' and 'usability improvement' together in seeing benefit from community input vs 'bug' (not really needing much discussion). This is I think where we somewhat differ in opinion. I would also say a number of items in the current feature request list could be considered usability improvements (Unmoved units finder, Streamline main menu UI, Add battle round to battle window, etc): https://forums.triplea-game.org/topic/182/guidelines-and-feature-request-list

Based on your categories, I'm not sure where improving an existing feature fits in (Ex. AI Bonus Settings Revamp). I would consider that a feature request though it isn't new, it was improving an existing feature.

I think its probably best to try and have just 2 main groups of items:
1. Those that have significant benefit from community input and discussion. (track in forum)
2. Those that don't need community input and discussion. (track in github)

I tend to put 'bugs' in 2 and everything else in 1. And I define bugs as being something not functioning or incorrect where fixing and how to fix it has little debate.

**Comment by DanVanAtta** [comment]
Thanks for the response discussion @ron-murhammer, glad we are taking some time to clarify this a bit.

I think we need to track active problems. Mixing these in forum threads I suspect would get them lost. I think a good example would be if someone were to report the many different UI spacing problems we have. That would be a good series of issues to work from, and IMO inappropriate for forum threads. 30 UI spacing issues I suspect would drown out every other topic. Maybe it would work okay, though it is not obvious to me that it would.

I think the balance is we need to 'ice-box' the things that we will never get to, and at the same time make sure we encourage testers to report a whole variety of issues and to report freely. If we see that 2 out of 3 problem reports are met with "this is valid, but it's not a bug", then it creates the incentive to not report.

Open for discussion, my current thoughts and suggestion would be:
- let's track all problems in github, 'sensitive' problems in admin forum
- anything that is minor or we will just never get to, or we never have to get to because is it is not a problem and does not hurt the game, 'ice-boxing' is perhaps the best thing to do. We should be honest about our priorities and what we want and will get fixed.


--------------------

That suggestion in context I think is interesting, for:
- Income Summary #2216, we would keep tracking it until someone picks it up. They would spend some time looking at the problem and ideally post a forum post with a proposal. Thinking about this, what's the value in sending this to the forum? We are adding work and tracking, there should be benefit. There is nobody right now willing or volunteering to fix it, I don't think it's even useful to invite feedback on the issue. If I were to pick it up, I would not want to be locked in to someone else's idea that I may not necessarily like or think workable.  Meanwhile, it is still a problem, I do not want us to sweep these under the rug, the cost of tracking is *not* that great. To some extent, are we just only going to track projects and major bugs? 

TripleA is interesting in that some features are really well done, and that I think is why the game is so successful. Though it does suffer from a number of half baked features, and we see that in the code with lots of dead code paths of things that people started adding but never finished. This latter part is a major weakness, I think we are at a point where we can and really should start tracking things that barely work as well as the things that don't work.


------------

Bring Combat Window to Front #2217

Note, there is actually a forum post already that is linked in the ticket. The use-case and feedback is potentially very varied, a wider audience I thought was pretty clearly useful. But by the philosophy that we should track problems, #2217 serves as a stub to remind us, this is a problem that should be addressed.

> There are probably a few others that could go either way and are maybe more feature/improvement than bug but would have limited community input value.

Summarizing, IMO track problems, cut forum threads when there is valuable feedback to be had. If something is not a problem, then it'll be a forum thread only (which to me matches the definition of a feature request nicely). 'usability' problems might often get forum threads, but I think it's premature and likely counter-productive even to cut one until someone is ready to follow through with a fix.

**Comment by DanVanAtta** [comment]
tl;dr I suggest:
- track bugs in github
- track usability problems/existing feature improvements in github, ice box them if open for some months. Let a dev picking up those ticket open forum threads as needed
- open forum threads for feature requests

**Comment by Cernelius** [comment]
The matter with bugs in forum, instead of in this Issue tracker, that it is obviously the right place to track bugs, as the sourceforge bug tracker was previously to GitHub, is not discussing the bug in forum, but the fact that my experience tells me that very few normal users report bugs in GitHub, even when instructed and directed to (by me).

To be clear, you are not getting bug reports in here from the normal users, but very rarely, even if the engine is full of bugs. Of course, just my perception.

Also, I realise you can get it clicking on help, but I would not have removed the link in the lobby chatlog for bug reporting nor the one for hosting information, and suggest to put them both back, right under Lobby Rules.

However, anyone can report bugs in GitHub (it is not hard), thus, if they don't want to, it's their choice not to report the bug.

On the matter at this topic, I think you should divide stuff on the line of tracking in forum everything that it is an addition or an effective change (if an improvement or not, it is part of the discussion), meaningfully effecting the user, of something that has no obvious issues, while tracking in here anything that needs additions or changes, and just should not be left as it is, plus anything that it is a merely display changes.

For example, if I want to propose renaming "Map Options" to "Game Options" (that is actually something wrong, because those options are the options for the game, not for the map, and can be totally diferrent for the games of the same map), that should be tracked in GitHub both because it is an actual issue (the name is wrong) and because it would be just silly to open a forum discussion about that.

In particular in the case of the usability, that would be a matter better kept in forum for discussion, actually, or maybe the best one, as it is the field where the users can have an idea the most, except that it is better tracking in GitHub usability issues, meaning stuff that it is wrong or has problems, thus should stay there open and tracked till it is solved.

**Comment by Cernelius** [comment]
If it would be up to me, on that latest bunch I would say:

.I don't understand:
6157 - Different staging screen treatments

.GitHub triplea-game:
6157 - no 'unmute' option
6157 - default player name for network games should be same as lobby player name
6157 - Missing step name pop-up in combat
6157 - combat button text is truncated, could use some hover text
6157 - 'quit' does not actually quit the game
6157 - WaW - sub in combat can still move, but unable to move it
6157 - download maps window resizing is not good
6157 - install folder is no longer versioned

.GitHub triplea-maps:
6157 - no 'over buy' warning on 270BC
then, since it is not a problem in 270BC, close and open in Forum a new topic about: 6157 - 'over buy' warning is not consistent, no user notification when it is turned on or off.

.Forum:
6157 - no longer have retreat confirmation when can only retreat to single territory
6157 - missing a min size on the game window
(I don't see it as a bug, but i don't understand what the first image is or means)
6157 - Bring to 'front' combat window button would be useful
6157 - PU 'end with' summary is overly verbose, calculations are not clear
(preferably in the already opened thread about this)

I think all at the first list is something that should be changed because it is an issue or an obvious improvement or missing item of something already available or generally a fix or something unrelated to playing the game as a normal user.
Instead, all at the last list are changes that are either added features or feature modifications in the usability that are not obvious.

But, in my opinion, the biggest problem is that it is needed to clarify and exactly explain how stuff works; then, it would be needed to be sure to close and redirect coherently.
Otherwise, if there is not a clear enough classification, the consequente is tracking stuff spuriously in two different places, and you have to search in both about the same things, which is the worst.

So, for example, once you decide on anything, then I think you should:
- make a wiki explaining clearly to anyone the process, about how it is decided what it is for forum and what it is for here, in a way that people might try to get it right, thus less need for you to keep closing stuff and redirect;
- close all issues that you believe should stay in forum and mandate reopening them in forum;
- revisit all the ice-box issues and eventually remove the label and reopen those that should stay here (in case you simply put the ice-box on anything having the feature request label).

Otherwise, in my opinion, it would be just confusing (as I'm still quite confused), and would be rather better tracking either everything in forum or everything in here.
So, please, once you decide the process make a wiki about it that it is easy for everyone to find, and also link it in forum, so people know where they are supposed to go, otherwise you get reports in forum that gets overlooked by the developers, thus wasting people time, or you have to keep closing and redirect while people randomly open here or in forum, without knowing where and what.

**Comment by prastle** [comment]
Yes I think Cer nailed it well with his explanation of bugs in forum.

  "The matter with bugs in forum, instead of in this Issue tracker, that it is obviously the right place to track bugs, as the sourceforge bug tracker was previously to GitHub, is not discussing the bug in forum, but the fact that my experience tells me that very few normal users report bugs in GitHub, even when instructed and directed to (by me).

To be clear, you are not getting bug reports in here from the normal users, but very rarely, even if the engine is full of bugs. Of course, just my perception."

He is correct.  This is why I suggested we try to report bugs in the Forum as well.  I do agree tho that once it has been reported the bug should be moved to git hub if this is possible somehow.

**Comment by panther2** [comment]
From my forum experience "bugs" reported there are often related to misunderstandings, wrong settings, problems with Java and the OS. So for these the forum is the perfect place to report, as regularly they can be resolved quickly. 

This is why I vote for Github as "real" bug tracker and the forum as additional platform to report.

**Comment by DanVanAtta** [comment]
Thanks for the feedback @Cernelius (and all). I suggest we continue to track all usability issues in github only.* We need a developer who is willing to own or fix any one of them before it is ever worth discussing as a larger group. This way we keep github as a single list of items that should not be left 'as-is'.

A key point then is that it is up to the developer the freedom to open a forum post when it is helpful. This way in a way resolves the duplication problem. The person fixing a usability issue and opens a forum post for wider feedback could cross link. Ideally we could assign that issue to that person, then we would know that the topic is being covered.  

*270BC included, that issue is not actually specific to 270BC but due to game engine hack that a feature was turned off completely to support custom map behavior at the cost of inconsitent engine behavior. Such an option is not without benefit, but we really should do better.


---------------


'ice-box' is a bit of an category. It's basically a gravestone marker, a list useful for anyone that is looking for something to work on. There is *no* value in moving those over to forum posts, it's a waste of time. The topics are also deemed honestly not something that we can't leave as-is, so leaving in github is not good either. Hence, it falls under the case of, if somebody is willing to do the work, then they can make the right call on when or if to open a forum post.

To be clear too, cowboy large scale changes will probably run into a lot of pain when we want to merge the branch. Forum posts or github issues are useful ways to coordinate and avoid that pain. In addition, we have a new 'beta' flag which we can put things under to allow us to do that after that fact, so we can open forum posts or github issues with instructions to turn on a flag and try out something that is a work in progress.

Since we don't want to open forum posts without an owner, and we want to close things in github that are not going to go anywhere, 'ice-box' is a bit of a stop-gap so we don't just outright close every issue - some we may want to keep tabs on in case there is new effort available. Hence it does not make sense to open 'ice-box' issues or create a duplicate forum post unless there is a developer willing to follow through.

**Comment by RoiEXLab** [comment]
Once we find a solution for this problem, we need to remeber to update the website accordingly...
This issue is opened for years now: https://github.com/triplea-game/triplea-game.github.io/issues/17

**Comment by DanVanAtta** [comment]
> Once we find a solution for this problem

Problems with long threads, we actually have...


------------------------------

tl;dr

Usability problems are problems, so we track in bug tracker. Developer that wants to pick up such a problem will create forum post when time is right. In contrast, features, just nice to have add-ons, go only in forums since they are not problems, so this way they do not pollute our bug and problem tracking queue.
